File: backend/solucao_fernando_emidio/Order.rb
#!/usr/bin/ruby -w

require "InvalidProductException"

class Order
  attr_reader :customer, :items, :payment, :address, :closed_at, :type
  
  def initialize(customer, type, overrides = {})
    @customer = customer
    @type = type
    @items = []
    @order_item_class = overrides.fetch(:item_class) { OrderItem }
    @address = overrides.fetch(:address) { Address.new(zipcode: '45678-979') }
  end
  
  def add_product(product)    
    if product.type !=  @type[:type]    
      raise InvalidProductException, "Impossível inserir um produto do tipo #{product.type} em uma ordem do tipo #{@type[:type]}"
    end
    
    @items << @order_item_class.new(order: self, product: product)
  end
  
  def total_amount
    @items.map(&:total).inject(:+)
  end
  
  def close(closed_at = Time.now)
    @closed_at = closed_at
  end
  
  def orderProductType
    @type
  end

  # remember: you can create new methods inside those classes to help you create a better design
end

File: backend/desafio_alynne/order.py

from address import Address
from customer import Customer
from product import Product
import time

class Order:

    customer = Customer

    items = [Product]

    deliveryAddress = Address

    closed_at = None

 

    def __init__(self, customer,deliveryAddress ):

        self.customer = customer

        self.items = [Product]

        self.deliveryAddress = Address(deliveryAddress)

 

    def add_product(self, product):
        
        self.items.append(product)

 

    def total_amount(self):

        total = 0

        for x in range(1, len(self.items)):
          total += self.items[x].value


        return total 

    def close(self, closed_at=time.time()):

        self.closed_at = closed_at

    def send_orders(self):
      # mais de um tipo d eproduto pode está associado ao mesmo produto
      for x in range(1, len(self.items)):
          print(self.items[x].productType)
          self.items[x].envia()
